daemon/disk_usage: Use context aware singleflight

vvoland opened the pull request:
Follow up to:
- https://github.com/moby/moby/pull/44365#discussion_r1007313872

Relates to:

- https://github.com/moby/moby/pull/42715 (it replaces the implementation introduced in that PR)

The singleflight function was capturing the context.Context of the first caller that invoked the `singleflight.Do`. So context cancellation of the first caller could cancel the execution of other concurrent invocations, even if their context wasn't cancelled.

singleflight calls were also moved up from the ImageService to Daemon. This allows to avoid duplicating this behaviour in every ImageService implementation (with containerd integration, we have two now).

vvoland commented on vendor.mod:
```diff
@@ -89,6 +89,7 @@ require (
 	google.golang.org/genproto v0.0.0-20220706185917-7780775163c4
 	google.golang.org/grpc v1.48.0
 	gotest.tools/v3 v3.4.0
+	resenje.org/singleflight v0.3.0
```
Are we fine with vendoring this dependency? It's not popular but it's small and I don't see anything widely used that could be used instead.

corhere commented on daemon/daemon.go:
```diff
@@ -62,10 +63,10 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 	"google.golang.org/grpc/credentials/insecure"
+	ctxsingleflight "resenje.org/singleflight"
```
[Avoid renaming imports except to avoid a name collision](https://github.com/golang/go/wiki/CodeReviewComments#imports)

vvoland replied:
Yeah I wasn't sure about it. My intent here is to make it explicitly visible, that this isn't the singleflight from `go/x/sync`.
On the other hand, it still does the same thing but with context handling, so it's probably fine not to disambiguate them.

corhere commented on daemon/daemon.go:
```diff
@@ -105,7 +106,10 @@ type Daemon struct {
 	seccompProfilePath string
 
-	usage singleflight.Group
+	usageContainers ctxsingleflight.Group[string, []*types.Container]
+	usageImages     ctxsingleflight.Group[string, []*types.ImageSummary]
+	usageVolumes    ctxsingleflight.Group[string, []*volume.Volume]
+	usageLayer      ctxsingleflight.Group[string, int64]
```
Given that only one constant "key" is used for each group respectively, I think it would make the code more self-documenting to use a key type with only one possible value: `struct{}`.

vvoland replied:
Good point, thanks!

corhere commented on daemon/disk_usage.go:
```diff
@@ -23,15 +26,55 @@ func (daemon *Daemon) ContainerDiskUsage(ctx context.Context) ([]*types.Containe
-			return nil, res.Err
-		}
-		return res.Val.([]*types.Container), nil
+	if err != nil {
+		return nil, err
 	}
+	return res, nil
```
```suggestion
	return res, err
```

vvoland replied:
Thanks, done!

corhere reviewed (commented):
> singleflight calls were also moved up from the ImageService to Daemon. This allows to avoid duplicating this behaviour in every ImageService implementation (with containerd integration, we have two now).

💯 Great call! In addition to DRYing the implementations, this change is also architecturally sound IMO.

vvoland committed:
daemon/disk_usage: Use context aware singleflight

The singleflight function was capturing the context.Context of the first
caller that invoked the `singleflight.Do`. This could cause all
concurrent calls to be cancelled when the first request is cancelled.

singleflight calls were also moved from the ImageService to Daemon, to
avoid having to implement this logic in both graphdriver and containerd
based image services.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere reviewed (approved):
LGTM!

thaJeztah commented:
Failure looks unrelated, but I don't recall seeing it before, so let me put it in a comment for posterity;

```
=== FAIL: github.com/docker/docker/integration-cli TestDockerAPISuite/TestContainersAPICreateMountsCreate/0_config:_{volume__c:\foo_false__<nil>_<nil>_<nil>_<nil>} (24.07s)
    docker_api_containers_test.go:2137: timeout hit after 10s: first check never completed
        --- FAIL: TestDockerAPISuite/TestContainersAPICreateMountsCreate/0_config:_{volume__c:\foo_false__<nil>_<nil>_<nil>_<nil>} (24.07s)
723
```

thaJeztah reviewed (approved):
LGTM! Thanks
